plugin/src/test/java/hudson/plugins/swarm/PipelineJobTest.java

```diff
@@ -0,0 +1,144 @@
+package hudson.plugins.swarm;
+
+import static org.junit.Assert.assertNotNull;
+
+import hudson.model.Node;
+import hudson.plugins.swarm.test.ProcessDestroyer;
+import hudson.plugins.swarm.test.TestUtils;
+import java.io.IOException;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep;
+import org.junit.Assume;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+public class PipelineJobTest {
+
+    @Rule public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @ClassRule public static TemporaryFolder tmpDir = new TemporaryFolder();
+
+    private final ProcessDestroyer processDestroyer = new ProcessDestroyer();
+
+    /** Executes a shell script build on a Swarm Client agent. */
+    @Test
+    public void buildShellScript() throws Exception {
+        story.then(
+                s -> {
+                    Node node =
+                            TestUtils.createSwarmClient(
+                                    story.j, processDestroyer, tmpDir.newFolder());
+
+                    WorkflowJob project = story.j.createProject(WorkflowJob.class);
+                    project.setConcurrentBuild(false);
+                    project.setDefinition(new CpsFlowDefinition(getFlow(node, 0), true));
+
+                    WorkflowRun build = story.j.buildAndAssertSuccess(project);
+                    story.j.assertLogContains("ON_SWARM_CLIENT=true", build);
+                    tearDown();
+                });
+    }
+
+    /**
+     * Executes a shell script build on a Swarm Client agent that has disconnected while the Jenkins
+     * master is still running.
+     */
+    @Test
+    public void buildShellScriptAfterDisconnect() throws Exception {
+        story.then(
+                s -> {
+                    Node node =
+                            TestUtils.createSwarmClient(
+                                    story.j, processDestroyer, tmpDir.newFolder());
+
+                    WorkflowJob project = story.j.createProject(WorkflowJob.class);
+                    project.setConcurrentBuild(false);
+                    project.setDefinition(new CpsFlowDefinition(getFlow(node, 1), true));
+
+                    WorkflowRun build = project.scheduleBuild2(0).waitForStart();
+                    SemaphoreStep.waitForStart("wait-0/1", build);
```

Comment:
Note that since this is waiting in `node` but not `sh` it only exercises part of https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/104 (when https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/104/commits/58b127fa34cd10bfc76659b86292832965fe7c95 was added).

Reply:
> Note that since this is waiting in `node` but not `sh` it only exercises part of [jenkinsci/workflow-durable-task-step-plugin#104](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/104)

Good point. I added a new test, `PipelineJobTest#buildShellScriptAcrossDisconnect`, which waits in `sh` (essentially a port of [`ExecutorStepTest#buildShellScriptAcrossDisconnect`](https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java#L322)).